### ISSUE TITLE
(Create "real" PR) Form validation prevents you from going "back" in a multi-page form

### ISSUE THREAD

**Issue by micahchiang** [department-of-veterans-affairs/formulate#87]
## Description
[Original DST ticket](https://app.zenhub.com/workspaces/vsp-design-system-5f8de67192551b0012ebb802/issues/department-of-veterans-affairs/va.gov-team/19473) (Opened 2/3/21)

Sometimes form validation prevents you from going "back" in a multi-page form. 

To reproduce: 
1. Start working on any benefit application
2. Move past the first page
3. Set focus on any required field
4. Then click the `back` button

![validate-error](https://user-images.githubusercontent.com/20728956/91746273-35859480-eb71-11ea-9df0-601a208ded1e.gif)

This also affects places where we are using `SchemaForm` outside of the form system, such as updating contact info in the Profile: 

1. Open an empty field to edit it
2. Focus a required field
3. Cancel to exit edit mode

You get a validation error and the form does not close:

![close.gif](https://images.zenhubusercontent.com/5b4d0309a8d8be5bda39303e/7ce46255-331f-402e-8891-4db987fc8433)

This issue is due to the form system's liveValidate feature (a prop on the Form component from @department-of-veterans-affairs/react-jsonschema-form), which validates the form whenever data changes. Specifically, each form input has an onBlur handler that triggers validation. When a user clicks the BACK or CANCEL button, the first thing that happens is the previously focused form input's onBlur handler fires. This triggers form validation. And that form validation somehow* short circuits the user's click on the BACK or CANCEL button. The button's click handler is never fired.

This issue was originally documented in https://github.com/department-of-veterans-affairs/va.gov-team/issues/9765.

And [this PR](https://github.com/department-of-veterans-affairs/vets-website/pull/14220) was submitted to fix it, but isn't an ideal solution and was never merged.

**Comment by micahchiang** [comment]
Komes could not reproduce

**Comment by micahchiang** [comment]
Other engineers to try and reproduce independently.

**Comment by bjmccotter7192** [comment]
I was able to recreate the issue described in the ticket in the staging environment while logged in as "Andrea" aka Vets Mock User 2.
When clicking into the field that is required and then trying to hit the back button, it forces me to have to put something in the field then allows me to go back.

**Comment by kkomes** [comment]
It turned out I tested on the different form. I tested it on the 10-10cg form, not the 10-10ez form.
 
Yes, I can now reproduce it on the 10-10ez. However, it worked fine on 10-10cg though.

**Comment by micahchiang** [comment]
I tested this on the 26-8832 and got a mixed result. Some fields prevented backward navigation while others did not.

**Comment by micahchiang** [comment]
@bjmccotter7192 to own this work for the rest of the sprint.

**Comment by bjmccotter7192** [comment]
Mob session notes and recorded upload located here [Mob Session 1/20](https://vfs.atlassian.net/wiki/spaces/FLT/pages/2061467674/Mob+Session+1+20+2022)
- Found issue in onBlur function on ProgressButton 
```onBlur(id) {
    if (!this.state.formContext.touched[id]) {
      const formContext = set(['touched', id], true, this.state.formContext);
      this.setState({ formContext });
    }
  }
```
This functions run instead of the onClick event on the Back/Continue Button

Proposed fix will be in branch `on-blur-back-button`

**Comment by bjmccotter7192** [comment]
checked in fix to ProgressButton.jsx adding the onMouseDown effect to trigger preventDefault to the buttons to bypass the onBlur touched functionality.
```
<button
        type={this.props.submitButton ? 'submit' : 'button'}
        disabled={this.props.disabled}
        className={`${this.props.buttonClass}${
          this.props.disabled ? ' usa-button-disabled' : ''
        }`}
        id={`${this.id}-continueButton`}
        onClick={this.props.onButtonClick}
        onMouseDown={e => {
          e.preventDefault();
        }}
        aria-label={this.props.ariaLabel || null}
        aria-describedby={this.props.ariaDescribedBy}
      >
        {beforeText}
        {this.props.buttonText}
        {afterText}
      </button>
```
This fix ensures that when you click the back button that the validation is ignored but when clicking on Continue the validation is ran. Also ensures that if you click anywhere on the page after selecting a required element that the onBlur function still acts as normal.

**Comment by bjmccotter7192** [comment]
Fix checked back into branch `on-blur-back-button` where I added a test for each scenario, created a `preventOnBlur` property on the component so it was more easily testable, and also moved the preventDefault out into a default property allowing the user to override it by passing their own function in (if they want to).

The tests test both scenarios of whether the property is passed to the component or not, also another test was added to make sure that the mouse down and click work together.

Checked in the branch if you would like to take a look at how I went about fixing the issues, going to create a PR soon.
